File: iteratorpattern/iteratorpattern.js
// Abstract Aggregate
class Aggregate {
  createIterator() {
    throw new Error('Method "createIterator()" must be implemented.');
  }
}

// Concrete Aggregate
class Library extends Aggregate {
  constructor(bookList) {
    super();
    this.bookList = bookList;
  }

  createIterator() {
    return new BookIterator(this.bookList);
  }
}

// Abstract Iterator
class Iterator {
  hasNext() {
    throw new Error('Method "hasNext()" must be implemented.');
  }

  next() {
    throw new Error('Method "next()" must be implemented.');
  }
}

// Concrete Iterator
class BookIterator extends Iterator {
  constructor(books) {
    super();
    this.books = books;
    this.index = 0;
  }

  hasNext() {
    return this.index < this.books.length;
  }

  next() {
    if (this.hasNext()) {
      const book = this.books[this.index];
      this.index += 1;
      return book;
    }
    return null;
  }
}

// Book class
class Book {
  constructor(price, bookName) {
    this.price = price;
    this.bookName = bookName;
  }

  getPrice() {
    return this.price;
  }

  getBookName() {
    return this.bookName;
  }
}

// Client code
const bookList = [
  new Book(100, 'Science'),
  new Book(200, 'Maths'),
  new Book(300, 'GK'),
  new Book(400, 'Drawing')
];

const lib = new Library(bookList);
const iterator = lib.createIterator();

while (iterator.hasNext()) {
  const book = iterator.next();
  console.log(book.getBookName());
}

File: iteratorpattern/iteratorpattern.py
from abc import ABC, abstractmethod

class Aggregate(ABC):
    @abstractmethod
    def create_iterator(self):
        pass

class Library(Aggregate):
    def __init__(self, book_list):
        self.book_list = book_list

    def create_iterator(self):
        return BookIterator(self.book_list)
    
class IteratorInterface(ABC):
    @abstractmethod
    def has_next(self):
        pass

    @abstractmethod
    def next(self):
        pass

class BookIterator(IteratorInterface):
    def __init__(self, books):
        self.books = books
        self.index = 0

    def has_next(self):
        return self.index < len(self.books)
    
    def next(self):
        if self.has_next():
            book = self.books[self.index]
            self.index += 1
            return book
        
        return None

class Book:
    def __init__(self, price, book_name):
        self.price = price
        self.book_name = book_name
    
    def get_price(self):
        return self.price
    
    def get_book_name(self):
        return self.book_name

if __name__ == '__main__':
    book_list = [
        Book(100, 'Science'),
        Book(200, 'Maths'),
        Book(300, 'GK'),
        Book(400, 'Drawing')
    ]

    lib = Library(book_list)
    iterator = lib.create_iterator()

    while iterator.has_next():
        book = iterator.next()
        print(book.get_book_name())
